File: apps/client-macos/Sources/TimeTrack/Notifications/FallbackDistractionNotifier.swift
import Foundation

/// Wraps the system notifier for the DISTRACTION nudge only. It posts a normal local notification
/// when macOS has authorized notifications, and falls back to a dismissible in-app window when it
/// has NOT (e.g. an un-notarized dev build the OS refuses to register) — so the nudge is never
/// silently dropped. Only the distraction path uses this; the idle / forgot-to-start / end-of-day
/// nudges keep the plain notifier and its documented silent-drop behaviour.
///
/// It conforms to `LocalNotifying` so `DistractionMonitor` stays unchanged and UI-free — the
/// monitor still just calls `notify(id:title:body:)` once per streak, and the authorized-vs-window
/// decision lives here. The strings it forwards are the generic, category-derived nudge title/body
/// the monitor produced — never an app name, host, or window title (CLAUDE.md §1).
final class FallbackDistractionNotifier: LocalNotifying {
    private let primary: LocalNotifying
    /// Async check of the current authorization state (real wiring reads
    /// `UNUserNotificationCenter.getNotificationSettings`; tests inject a synchronous stub).
    private let isAuthorized: (@escaping (Bool) -> Void) -> Void
    /// Presents the in-app fallback window. Real wiring hops to the main thread; kept as a plain
    /// closure so the decision logic is testable without AppKit.
    private let presentWindow: (_ title: String, _ body: String) -> Void

    init(primary: LocalNotifying,
         isAuthorized: @escaping (@escaping (Bool) -> Void) -> Void,
         presentWindow: @escaping (_ title: String, _ body: String) -> Void) {
        self.primary = primary
        self.isAuthorized = isAuthorized
        self.presentWindow = presentWindow
    }

    func requestAuthorization() { primary.requestAuthorization() }
    func clearAll() { primary.clearAll() }

    func notify(id: String, title: String, body: String) {
        isAuthorized { [weak self] authorized in
            guard let self else { return }
            if authorized {
                self.primary.notify(id: id, title: title, body: body)
            } else {
                self.presentWindow(title, body)
            }
        }
    }
}

File: apps/client-windows/src/NiftyTimer/Notifications/FallbackDistractionNotifier.cs
namespace NiftyTimer.Notifications;

/// <summary>
/// Wraps the notifier for the DISTRACTION nudge only. Ported from the macOS client's
/// <c>FallbackDistractionNotifier</c>: a normal notification when Windows will show one, and a
/// dismissible in-app card when notifications are switched off, so the nudge is never silently
/// lost. The idle, forgot-to-start and end-of-day nudges keep the plain notifier and its
/// documented silent drop.
///
/// It is an <see cref="ILocalNotifier"/>, so <see cref="Tracking.DistractionMonitor"/> stays
/// unchanged and UI-free. It forwards only the generic, category-derived text the monitor produced
/// — never an app name, host or window title (CLAUDE.md §1).
/// </summary>
public sealed class FallbackDistractionNotifier : ILocalNotifier
{
    private readonly ILocalNotifier _primary;
    private readonly Func<bool> _notificationsEnabled;
    private readonly Action<string, string> _presentWindow;

    public FallbackDistractionNotifier(
        ILocalNotifier primary,
        Func<bool> notificationsEnabled,
        Action<string, string> presentWindow)
    {
        _primary = primary;
        _notificationsEnabled = notificationsEnabled;
        _presentWindow = presentWindow;
    }

    /// <summary>
    /// Asked on every nudge rather than once: switching notifications back on takes effect at the
    /// very next one.
    /// </summary>
    public void Notify(string id, string title, string body)
    {
        if (_notificationsEnabled())
        {
            _primary.Notify(id, title, body);
        }
        else
        {
            _presentWindow(title, body);
        }
    }
}
